### ISSUE TITLE
extend scanner.Definition to support subpatterns/subdefinitions

### ISSUE THREAD

**Issue by alaska** [sourcegraph/scanner#1]
I'm working on using sourcegraph/scanner and sourcegraph/lexer to create a nice pure go solution for syntax highlighting.  Currently the libraries meet most of my needs, but I feel a couple of additions could make things easier to add nicer functionality and support more languages.

Specifically, I think it would be nice to extend scanner.Definition to support subpatterns.  Something like this:

```go
type Definition struct {
    Regex *regexp.Regexp
    Type string
    subpatterns []Definition
}

func (d Definition) Contains( s ...Definition) Definition {
    d.subpatterns = append(d.subpatterns, s...)
    return d
}
```

This would allow use of the patterns in lexer/common.go, but extending them in a language-specific way to support nice little additions like highlighting verbs and escapes in strings:

```go
//In lexer/languages.go:
Languages = map[string]*Language{
"Go": &Language{
    Extensions: []string{"go"},
    Map: append([]scanner.Definition{
        // Single line comments
        regex["comments"]["oneline"],

        // Multi-line comments
        regex["comments"]["multiline"],

        // Double quote strings
        regex["strings"]["double"].Contains(
            scanner.Definition{regexp.MustCompile(`%[#0\-[:space:]+*]*[vTtbcdoqxXUeEfgGsp]`), "VERB"},
        ),
        // ...
```
Any thoughts on whether something like this would make a welcome addition?

**Comment by shurcooL** [comment]
> I'm working on using sourcegraph/scanner and sourcegraph/lexer to create a nice pure go solution for syntax highlighting.

Hi, since I'm not familiar with your work yet, I want to ask this first:

Are you aware of https://github.com/sourcegraph/syntaxhighlight and https://github.com/sourcegraph/annotate packages? If so, are they not sufficient, and how would your syntax highlighting solution differ?

I'm just looking to clarify. Thanks!

**Comment by alaska** [comment]
I actually have a [solution](https://github.com/alaska/hugo/commit/3942b19320d6097fb11e69baa558e242bd8b3856) in place for my initial use case (removing the JS/Pygments dependencies from Hugo) that uses https://github.com/sourcegraph/syntaxhighlight, but it seems like it's mostly geared towards being language-agnostic for a select subset of c-like languages, and I would at the very least like  to support HTML and maybe even something like Perl at some point.

I wasn't aware of annotate.  Is that the "language-independent lexer" that syntaxhighlight uses?

Mostly what I'm after is a layer on top of something that generates nicely-formatted HTML (eventually extensible to TeX/Postscript/etc) with some additional features like in-string highlight of verbs and escape sequences and tabular views of code supporting line numbers and highlightable lines/blocks. As such, I'm not interested in changing the base library too drastically, and in fact I already have a working POC for the in-string highlighting, but it would certainly be easier if subpatterns were supported in the library.

The main thrust of this issue is whether or not this kind of thing is even the focus of lexer/scanner.

**Comment by shurcooL** [comment]
> https://github.com/sourcegraph/syntaxhighlight, but it seems like it's mostly geared towards being language-agnostic for a select subset of c-like languages

Correct, `syntaxhighlight` is a general purpose highlighter for C-like (and other) languages. It does a pretty good job of general highlighting, but it's not perfect for any specific language.

> I wasn't aware of annotate. Is that the "language-independent lexer" that syntaxhighlight uses?

Annotate allows applying multiple sets of annotations to text. For example, code highlighting in addition to some sort of other markup (perhaps this is similar to what you're describing as "in-string highlight of verbs"?).

I've also written a [`github.com/shurcooL/go/highlight_go`](https://godoc.org/github.com/shurcooL/go/highlight_go) package that is similar to `syntaxhighlight`, but targets Go specifically (and only Go). It uses `go/scanner` internally. Just sharing in case you find it relevant.
